pinot-core/src/main/java/org/apache/pinot/core/segment/processing/genericrow/GenericRowSerializer.java

```diff
@@ -0,0 +1,240 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.processing.genericrow;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+
+/**
+ * Utility class to serialize the {@link GenericRow}.
+ * The bytes are stored in NATIVE order. The data should be deserialized by the {@link GenericRowDeserializer} on the
+ * same host to ensure that both of them are using the same byte order.
+ */
+public class GenericRowSerializer {
+  private final int _numFields;
+  private final String[] _fieldNames;
+  private final boolean[] _isSingleValueFields;
+  private final DataType[] _storedTypes;
+  // Cache the encoded string bytes
+  private final Object[] _stringBytes;
+  // Store index for null fields
+  private final Map<String, Integer> _fieldIndexMap;
+  private final int[] _nullFieldIndexes;
+
+  public GenericRowSerializer(List<FieldSpec> fieldSpecs, boolean includeNullFields) {
+    _numFields = fieldSpecs.size();
+    _fieldNames = new String[_numFields];
+    _isSingleValueFields = new boolean[_numFields];
+    _storedTypes = new DataType[_numFields];
+    _stringBytes = new Object[_numFields];
+    for (int i = 0; i < _numFields; i++) {
+      FieldSpec fieldSpec = fieldSpecs.get(i);
+      _fieldNames[i] = fieldSpec.getName();
+      _isSingleValueFields[i] = fieldSpec.isSingleValueField();
+      _storedTypes[i] = fieldSpec.getDataType().getStoredType();
+    }
+    if (includeNullFields) {
+      _fieldIndexMap = new HashMap<>();
+      for (int i = 0; i < _numFields; i++) {
+        _fieldIndexMap.put(_fieldNames[i], i);
+      }
+      _nullFieldIndexes = new int[_numFields];
+    } else {
+      _fieldIndexMap = null;
+      _nullFieldIndexes = null;
+    }
+  }
+
+  /**
+   * Serializes the given {@link GenericRow}.
+   */
+  public byte[] serialize(GenericRow row) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 111 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 159-236 into a separate method.